specification.md

```diff
@@ -0,0 +1,501 @@
+
+## The Data Model
+
+Based on our research and avoiding a lot of the jargon, there are the following “components” in all the single-cell projects:
+
+* One or more **labeled 2D arrays** that we will call `X` (similar to AnnData and CXG - Seurat and SingleCellExperiment call them `assays`, and Loom calls them `layers`).
+* A **dataframe** called `obs`, which contains the row labels for the `X` arrays (in SingleCellExperient the `X` arrays are transposed, so `obs` are the column labels).
+* A **dataframe** called `var`, which contains the column labels for the `X` arrays (in SingleCellExperient the `X` arrays are transposed, so `var` are the row labels).
+* Any **key-value metadata**.
+* Any other number of auxiliary **dataframes** and **arrays**.
+
+In addition to the core model above, there exist various means of representing filtered & transformed datasets, and associating data from different assays/modalities.  We propose modelling as:
+
+* A **sc_group**, composed of the above elements -- a self-consistent set of `X` matrices (aka _layers_) and their axis labels (`obs`,`var`), where all three share consistent shape and indexing labels.  Every `sc_group` exists within a `sc_dataset`.
+* A **sc_dataset**, which is a set of named `sc_groups`.  Primary use is representing a collection of related annotated matrix data, including progressively filtered data (raw vs norm), related assays with different labels/shape, etc.
+
+Notes:
+
+* A “labeled 2D array” is just a 2D array that can be sliced potentially using string values, or values of any data type other than the typical integer matrix indices.
+* The way a labeled 2D array is implemented is _specific to the storage engine that implements it_. For example:
+    * It could be **dense**, in which case there must exist some kind of a mapping between the string labels and the integer indices.
+    * It could be **sparse**, in which case the labels could just be the string dimensions.
+* The auxiliary dataframes and arrays can be anything (e.g., `reducedDims` in SingleCellExperiment). These should be able to be efficiently joined with `obs` and `var` to produce the labels to query `X` with.
+* There can be more than one `X` arrays, and we will distinguish between two cases:
+    * They _share_ the same labels/coordinates and non-empty cells. In that case, we will talk about “layers” and use an intuitive API (see “attributes” in the TileDB implementation).
+    * They _do not share_ the same labels/coordinates. In that case, we will talk about separate arrays with their own coordinates.
```

Comment:
Recommend that we reword this (and the previous bullet) based on confusion in the gdoc -- the distinction is sharing the same non-empty cells (ie, the same coordinates are empty/non-empty).